Spec spoilers and color attribute allowance (#2549)
matrix-org/matrix-doc: specification/modules/instant_messaging.rst

:``font``:

>   ``data-mx-bg-color``, ``data-mx-color``, ``color``

Reviewer: Any chance the spec could recommend which of data-mx-color or color should be generated by new code? AFAICT color is better supported?

Reviewer 2: Wouldn't that need another MSC?

Reviewer: fair enough, apologies for being unfamiliar with the MSC process. should i suggest this as a new issue?

Reviewer 2: or you could also make a mini-msc for this. It should be really straight-forward, small and easy

For ``m.image``, the text should be ``"sent an image."``. For ``m.video``, the text
should be ``"sent a video."``. For ``m.audio``, the text should be ``"sent an audio file"``.

> Spoiler messages

Author: TODO: Mention that spoilers are only possible on `m.room.message` (specifically `m.text`) events.

If another MSC does show up, it'd be good to expand this to textual types in general (emotes, notices, etc).

~~~~~~~~~~~~~~~~

Parts of a message can be hidden visually from the user through use of spoilers. This does
not affect the server's representation of the event content - it is simply a visual cue to
the user that the message may reveal important information about something, spoiling any
relevant surprise.

Spoilers require use of the ``formatted_body`` and therefore the ``org.matrix.custom.html``

> message format, described above.

Reviewer 3: Strictly speaking, they don't; not for displaying, at least. We might probably mandate that clients that do support sending spoilers MUST send the rich-text version as well; but plain-text clients should not be deprived of an opportunity to save the surprise for their users.

Reviewer 2: uuuuuuh, how can you send spoilers without a `formatted_body`. Soru doesn't recall anything in the MSC on that. There was merely a *suggestion* on how the plaintext fallback *could* look like

Reviewer 3: I'm recommending the opposite: clients should send both representations but are allowed to process and display any of them, not necessarily the rich one.

Author: The proposal says this:

> This proposal is about adding a new attribute to the `formatted_body` of messages with type
`m.room.message` and msgtype `m.text`.

Which makes it a requirement that HTML be supported. The fallback support is left as an implied requirement, as it is in the MSC, though the file upload is a strong recommendation.

This PR also doesn't seem to mention that it only applies to `m.text` msgtypes, so will have to add that in.

Changing the HTML requirement feels like it should go through an MSC as it's a fairly substantial change to the original MSC, given the original MSC is based completely off the HTML representation.

Reviewer 3: > Which makes it a requirement that HTML be supported.

I see it the other way around: not "if the client wants to send spoilers it MUST support HTML" but "assuming the client supports HTML, it MUST use this convention to send/receive spoilers and SHOULD send the plaintext fallback". Again, I believe it's a different story for sending and receiving, and the original proposal does not rule out using the fallback to display spoilers (otherwise why even provide the fallback). Naturally (and I already mentioned it above), _sending_ spoilers implies using HTML.

Reviewer 2: The point of `body` is that clients who don't want to support formatting can just print that without having to format anything? Requiring clients to now parse the `body` for spoilers doesn't seem right to that? And yes, this seems like it would have been an appropriate discussion before the MSC was merged.

Reviewer 2: Currently when sending replies the spec also requires sending a `formatted_body`, btw

Reviewer 3: Replies as they are specced now are a perfect example of overengineering which should not be repeated though.

I guess I'll just propose a rewrite of this part (still in line with the original MSC) rather than try to explain what I'd like to see.

> Spoilers require use of the ``formatted_body`` and therefore the ``org.matrix.custom.html``
> message format, described above.

Reviewer 3: ```suggestion
To send spoilers clients MUST use the ``formatted_body`` and therefore the 
``org.matrix.custom.html``  format, described above.
```

Spoilers themselves are contained within ``span`` tags, with the reason (optionally) being
in the ``data-mx-spoiler`` attribute. Spoilers without a reason must at least specify the
attribute, though the value may be empty/undefined.

An example of a spoiler is:

.. code-block:: json

  {
    "msgtype": "m.text",
    "format": "org.matrix.custom.html",
    "body": "Alice [Spoiler](mxc://example.org/abc123) in the movie.",
    "formatted_body": "Alice <span data-mx-spoiler>lived happily ever after</span> in the movie."
  }

If a reason were to be supplied, it would look like:

.. code-block:: json

  {
    "msgtype": "m.text",
    "format": "org.matrix.custom.html",
    "body": "Alice [Spoiler for health of Alice](mxc://example.org/abc123) in the movie.",

>     "formatted_body": "Alice <span data-mx-spoiler="health of alice">lived happily ever after</span> in the movie."

Reviewer 2: ```suggestion
    "formatted_body": "Alice <span data-mx-spoiler=\"health of alice\">lived happily ever after</span> in the movie."
```

  }

> The plain text (``body``) fallback for spoilers is a little different than the HTML-formatted
> message as the ``body`` is often included as-is in notifications to the user. To prevent spoilers
> in notifications and other places, clients are strongly encouraged to first upload the spoiler

Reviewer 3: ```suggestion
When sending a spoiler, clients SHOULD provide the plain text fallback in ``body``.
The fallback SHOULD omit the spoiler text verbatim (but include the reason), since ``body``
contents can be used as-is in basic clients, notifications to the user etc. To prevent spoilers
showing up in such situations, clients are strongly encouraged to first upload the spoiler
```

Author: The MSC does not make the fallback a SHOULD, anywhere (or I'm failing to see it). The MSC only ever references it as "the plaintext fallback" in a context of SHOULD upload the text, implying that it is required (in terms of language on the page).

To be clear, this concern is about the first SHOULD - the second SHOULD with respect to uploading the files is fine.

I'm happy to block this PR behind another MSC to clarify the intent, as this seems like it could be an opportunity for extended discussion which doesn't belong here.

Reviewer 3: Yeah, I agree the original MSC is not perfectly clear on the modality of the fallback (whether it's SHOULD or MUST).

Reviewer 2: Soru intended to word it as a recommendation, however if it is ambiguous it seems like further discussion might be appropriate.

Author: https://github.com/matrix-org/matrix-doc/pull/2557
